components/nonmem/src/metadata.rs

```diff
@@ -0,0 +1,224 @@
+use anyhow::{Result, anyhow, bail};
+use fs_err as fs;
+use serde::{Deserialize, Serialize};
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+
+use utils::write_json_to_file;
+
+pub const METADATA_FILENAME_SUFFIX: &str = "_metadata.json";
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default, Hash, PartialEq, Eq)]
+pub struct ModelMetadata {
+    /// Parent model(s) this model is based on
+    #[serde(default)]
+    pub based_on: Vec<String>,
+    /// Short description of the model
+    pub description: String,
+    pub tags: Vec<String>,
+}
+
+impl ModelMetadata {
+    pub fn new(based_on: Vec<String>) -> Self {
+        Self {
+            based_on,
+            description: String::new(),
+            tags: Vec::new(),
+        }
+    }
+
+    pub fn load(path: impl AsRef<Path>) -> Result<Self> {
+        let content = fs::read_to_string(path)?;
+        Ok(serde_json::from_str(&content)?)
+    }
+
+    pub fn save(&self, model_name: &str, folder: impl AsRef<Path>) -> Result<()> {
+        let metadata_path = folder
+            .as_ref()
+            .join(format!("{model_name}{METADATA_FILENAME_SUFFIX}"));
+        write_json_to_file(self, metadata_path)?;
+        Ok(())
+    }
+
+    pub fn update(
+        mut self,
+        description: Option<String>,
+        mut tags: Vec<String>,
+        mut based_on: Vec<String>,
+        overwrite: bool,
+    ) -> Self {
+        if overwrite {
+            if let Some(d) = description {
+                self.description = d;
+            };
+            if !tags.is_empty() {
+                self.tags = tags;
+            };
+            if !based_on.is_empty() {
+                self.based_on = based_on;
+            };
+
+            return self;
+        }
+        // Append Tags
+        if !tags.is_empty() {
+            self.tags.append(&mut tags);
+        }
```

Comment:
I'm not sure what the right semantics are. Why is it appending vs replacing? How do you completely replace the tags?